File: leetcode-algorithms/141. Linked List Cycle/Solution.cpp
// Given a linked list, determine if it has a cycle in it.
//
// Follow up:
// Can you solve it without using extra space?
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if ((head == nullptr) || (head->next == nullptr)) return false;
        ListNode* speed1 = head;
        ListNode* speed2 = head;

        do {
            speed1 = speed1->next;
            speed2 = speed2->next->next;
            if (speed1 == speed2) return true;
        } while ((speed2 != nullptr) && (speed2->next != nullptr));

        return false;
    }
};

File: leetcode-algorithms/141. Linked List Cycle/Solution.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        if ((head == null) || (head.next == null)) return false;

        ListNode p1 = head;
        ListNode p2 = head;

        do {
            p1 = p1.next;
            p2 = p2.next.next;
            if (p1 == p2) return true;
        } while ((p1 != null) && (p2 != null) && (p2.next != null));

        return false;
    }
}
